### PR TITLE
Use SHA2 and PSS padding when creating client assertion

### LINKED ISSUE #4428: [Feature Request] AAD client assertions should be computed using SHA 256 and an approved padding scheme
### MSAL client type

Confidential

### Problem Statement

When MSAL creates the client assertion, it uses PKCS1 padding for digital signature and SHA1 as x5t claim. These are old crypto algorithms and we need to move to newer versions. The STS is building support.

See ESTS work items :

https://identitydivision.visualstudio.com/Engineering/_workitems/edit/2655345
https://identitydivision.visualstudio.com/Engineering/_workitems/edit/2704466

### Proposed solution

Use x5t#s256 and PSS padding when talking to ESTS, CIAM, B2C(?) but not with ADFS.

### PR DESCRIPTION
**Pull request by bgavrilMS**
Fixes #4428 

**Changes proposed in this request**
- use SHA2 and PSS for all authority types except Generic, dSTS and ADFS.
- refactor to use string operations instead of JSON serialization for creating assertions

**Testing**

**Performance impact**

Using strings operations instead of JSON serialization leads to a small ~1KB memory improvement while the performance is marginally better, ~10% gain.

### OLD
| Method          | UseX5C | UseExtraClaims | Mean     | Gen0   | Allocated |
|---------------- |------- |--------------- |---------:|-------:|----------:|
| SimpleAssertion | False  | False          | 690.4 us | 0.9766 |      8 KB |
| SimpleAssertion | False  | True           | 683.2 us | 0.9766 |  10.35 KB |
| SimpleAssertion | True   | False          | 680.3 us | 3.9063 |  25.43 KB |
| SimpleAssertion | True   | True           | 705.1 us | 3.9063 |  27.27 KB |

### NEW 
Rows 5-8 use SHA2

| Method          | UseSha2 | UseX5C | UseExtraClaims | Mean     | Gen0   | Allocated |
|---------------- |-------- |------- |--------------- |---------:|-------:|----------:|
| SimpleAssertion | False   | False  | False          | 687.2 us |      - |   7.06 KB |
| SimpleAssertion | False   | False  | True           | 658.0 us | 0.9766 |   9.08 KB |
| SimpleAssertion | False   | True   | False          | 640.8 us | 2.9297 |  23.56 KB |
| SimpleAssertion | False   | True   | True           | 657.4 us | 3.9063 |  25.59 KB |
| SimpleAssertion | True    | False  | False          | 662.6 us | 0.9766 |   8.27 KB |
| SimpleAssertion | True    | False  | True           | 657.8 us | 0.9766 |  10.31 KB |
| SimpleAssertion | True    | True   | False          | 670.3 us | 3.9063 |  24.78 KB |
| SimpleAssertion | True    | True   | True           | 678.4 us | 3.9063 |   26.8 KB |

**Documentation**
- [ ] All relevant documentation is updated.

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs [hunk 1]
```diff
@@ -325,6 +325,12 @@ internal bool IsDefaultAuthority
             }
         }
 
+        /// <summary>
+        /// True if SHA2 and PSS can be used for creating the client credential from a certificate 
+        /// </summary>
+        public bool IsSha2CredentialSupported => 
+            AuthorityType == AuthorityType.Aad;
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs:332]: TODO: test the other authorities

File: src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs [hunk 2]
```diff
@@ -131,45 +187,6 @@ private string Encode(X509Certificate2 certificate, string base64EncodedThumbpri
             return string.Concat(encodedHeader, ".", encodedPayload);
         }
 
-        [JsonObject]
-        [Preserve(AllMembers = true)]
-        internal class JWTHeader
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs:136]: This flow is too slow, as it relies on JSON serialization. But in fact the JSON is just a string.

File: src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs [hunk 3]
```diff
@@ -78,10 +86,55 @@ private static string UrlEncodeSegment(byte[] segment)
...
+            string header;
+
+            if (sendX5C)
+            {
+#if NETFRAMEWORK
+                string x5cValue = Convert.ToBase64String(certificate.GetRawCertData());
+#else
+                string x5cValue = Convert.ToBase64String(certificate.RawData);
+#endif
+                header = $$"""{"alg":"{{alg}}","typ":"JWT","{{thumbprintKey}}":"{{thumbprint}}","x5c":"{{x5cValue}}"}""";
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs:104]: This is much faster than creating an Object with these properties and applying JSON serialization.

File: tests/Microsoft.Identity.Test.Common/Core/Mocks/MockHttpManager.cs [hunk 4]
```diff
@@ -71,6 +72,15 @@ public MockHttpMessageHandler AddMockHandler(MockHttpMessageHandler handler)
 
         public int QueueSize => _httpMessageHandlerQueue.Count;
 
+        /// <summary>
+        /// For use only in threads that spin many threads. Not thread safe.
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Common/Core/Mocks/MockHttpManager.cs:76]:
```suggestion
        /// For use only in tests that spin many threads. Not thread safe.
```

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ClientCredentialsTests.cs [hunk 5]
```diff
@@ -81,10 +80,10 @@ public async Task WithSecret_TestAsync(Cloud cloud, TargetFrameworks runOn)
 
         [TestMethod]
         [DataRow(Cloud.Public, TargetFrameworks.NetFx | TargetFrameworks.NetCore)]
-        [DataRow(Cloud.Adfs, TargetFrameworks.NetFx)]
+        [DataRow(Cloud.Adfs, TargetFrameworks.NetFx | TargetFrameworks.NetCore)]
         [DataRow(Cloud.PPE, TargetFrameworks.NetCore)]
         // [DataRow(Cloud.Arlington)] - cert not setup
-        public async Task WithClientAssertion_Manual_TestAsync(Cloud cloud, TargetFrameworks runOn)
+        public async Task WithClientAssertion_AAD_Manual_TestAsync(Cloud cloud, TargetFrameworks runOn)
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ClientCredentialsTests.cs:86]:
```suggestion
        public async Task WithClientAssertion_Manual_TestAsync(Cloud cloud, TargetFrameworks runOn)
```

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ClientCredentialsTests.cs [hunk 6]
```diff
@@ -69,8 +68,8 @@ public async Task WithCertificate_TestAsync(Cloud cloud, TargetFrameworks runOn,
         }
 
         [TestMethod]
-        //[DataRow(Cloud.Public, TargetFrameworks.NetFx | TargetFrameworks.NetCore)]
-        //[DataRow(Cloud.Adfs, TargetFrameworks.NetFx)]
+        [DataRow(Cloud.Public, TargetFrameworks.NetFx | TargetFrameworks.NetCore)]
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ClientCredentialsTests.cs:71]: Not sure why we have disabled tests...

File: src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs [hunk 7]
```diff
@@ -303,6 +311,8 @@ internal bool IsDefaultAuthority
             }
         }
 
+
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs:314]: nit: extra lines

File: src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs [hunk 8]
```diff
@@ -3,273 +3,183 @@
...
-                {
-                    // This setter is required by the serializer
-                }
+#if NETFRAMEWORK
+                string x5cValue = Convert.ToBase64String(certificate.GetRawCertData());
+#else
+                string x5cValue = Convert.ToBase64String(certificate.RawData);
+#endif
+                header = $$"""
+{
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs:122]: needs proper spacing

File: src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs [hunk 9]
```diff
@@ -3,273 +3,183 @@
...
-                    return Certificate == null
-                        ? JsonWebTokenConstants.Algorithms.None
-                        : JsonWebTokenConstants.Algorithms.RsaSha256;
-                }
 
-                set
-                {
-                    // This setter is required by the serializer
-                }
+                header = $$"""
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs:133]: needs proper spacing

File: src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs [hunk 10]
```diff
@@ -3,273 +3,183 @@
...
 
-        private static string UrlEncodeSegment(byte[] segment)
-        {
-            return Base64UrlHelpers.Encode(segment);
-        }
+            int i = 0;
+            foreach (var kvp in _claimsToSign)
+            {
+                payload.Append($"\"{kvp.Key}\":\"{kvp.Value}\"");
+                if (i!= _claimsToSign.Count-1)
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs:72]: might be cleaner to do 
```
if (kvp.Key != _claimsToSign.Last.Key)
{
    payload.Append(",");
}
```
Then you can get rid of the i counter.

File: tests/Microsoft.Identity.Test.Performance/ClientAssertionTests.cs [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+namespace Microsoft.Identity.Test.Performance
+{
+    public class ClientAssertionTests
+    {
+        private static X509Certificate2 s_certificate = 
+            CertificateHelper.CreateCertificate("CN=rsa2048", RSA.Create(2048), HashAlgorithmName.SHA256, null);
+
+        string base64EncodedThumbprint = Base64UrlHelpers.Encode(s_certificate.GetCertHash());
+
```
Comment (trwalke) [inline tests/Microsoft.Identity.Test.Performance/ClientAssertionTests.cs:21]: extra line

File: tests/Microsoft.Identity.Test.Performance/ClientAssertionTests.cs [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+        private static Dictionary<string, string> s_cl = new Dictionary<string, string>()
+        {
+            {"key1", "val1" },
+            {"key2", "val2" }
+        };
+
+        //[ParamsAllValues]
```
Comment (trwalke) [inline tests/Microsoft.Identity.Test.Performance/ClientAssertionTests.cs:29]: dead code?

**Review by trwalke — APPROVED** [review]

File: src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs [hunk 13]
```diff
@@ -3,273 +3,183 @@
...
-            return UrlEncodeSegment(Encoding.UTF8.GetBytes(segment));
-        }
+            if (_appendDefaultClaims)
+            {
+                string defaultClaims = $$"""{"aud":"{{_audience}}","iss":"{{_clientId}}","sub":"{{_clientId}}","nbf":"{{validFrom}}","exp":"{{validTo}}","jti":"{{Guid.NewGuid()}}",""";
+                payload.Append(defaultClaims);
+            }
+            else
+            {
+                payload.Append("{");
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs:65]: Use Append(char) overload.

File: src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs (hunk 10 above)
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs:72]: Last is part of Linq, and I think in this case it'll lead to the enumeration of the entire collection.

File: src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs (hunk 13 above)
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs:65]: Done and added rule to .editorconfig

**Pull request closed without merging**